topology/probes/netlink/sriov.go

```diff
@@ -0,0 +1,208 @@
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/skydive-project/skydive/graffiti/graph"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology"
+)
+
+type pendingVf struct {
+	intf *graph.Node
+	vfid int
+}
+
+// PciFromString transforms the symbolic representation of a pci addres
+// in an unsigned integer.
+func PciFromString(businfo string) (uint32, error) {
+	var domain, bus, device, function uint32
+	_, err := fmt.Sscanf(
+		businfo,
+		"%04x:%02x:%02x.%01x",
+		&domain, &bus, &device, &function)
+	if err != nil {
+		return 0, err
+	}
+	return ((domain << 16) | (bus << 8) | (device << 3) | function), nil
+}
+
+// PciToString transforms un unsigned integer representing a pci address in
+// the symbolic representation as used by the Linux kernel.
+func PciToString(address uint32) string {
+	return fmt.Sprintf(
+		"%04x:%02x:%02x.%01x",
+		address>>16,
+		(address>>8)&0xff,
+		(address>>3)&0x1f,
+		address&0x7)
+}
+
+/* readIntFile reads a file containing an integer. This function targets
+specifically files from /sys or /proc */
+func readIntFile(path string) (int, error) {
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		return -1, err
+	}
+	result, err := strconv.Atoi(strings.TrimSpace(string(contents)))
+	if err != nil {
+		return -1, err
+	}
+	return result, nil
+}
+
+/* ProcessNode, the action associated to a pendingVf connects the node of actual
+   virtual to the VF node associated to the physical interface */
+func (pending *pendingVf) ProcessNode(g *graph.Graph, node *graph.Node) bool {
+	tr := g.StartMetadataTransaction(node)
+	tr.AddMetadata("VfID", int64(pending.vfid))
+	if err := tr.Commit(); err != nil {
+		logging.GetLogger().Errorf("Metadata transaction failed: %s", err)
+	}
+	if !topology.HaveLink(g, pending.intf, node, "vf") {
+		if _, err := topology.AddLink(g, pending.intf, node, "vf", nil); err != nil {
+			logging.GetLogger().Error(err)
+		}
+	}
+	return false
+}
+
+/* handleSriov adds a node for each virtual function declared. It takes
+   care of finding the PCI address of each VF */
+func (u *NetNsProbe) handleSriov(
+	graph *graph.Graph,
+	intf *graph.Node,
+	id int,
+	businfo string,
+	attrsVfs []netlink.VfInfo,
+	name string,
+) {
+	var pciaddress uint32
+	var offset, stride int
+	var link netlink.Link
+	var err error
+
+	// Yes we have to wait. May be considered as a race bug in Linux kernel
+	numVfsFile := fmt.Sprintf(
+		"/sys/bus/pci/devices/%s/sriov_numvfs", businfo)
+
+	if _, err = os.Stat(numVfsFile); os.IsNotExist(err) {
+		// There is no VFS to monitor.
+		return
+	}
+	count, numVfs := 0, 0
+	for numVfs == 0 && count < 5 {
+		numVfs, err = readIntFile(numVfsFile)
+		time.Sleep(1000 * time.Millisecond)
+		if err != nil {
+			logging.GetLogger().Errorf(
+				"SR-IOV: cannot get numvfs of PCI - %s", err)
+			return
+		}
+		count = count + 1
+	}
+
+	count = 0
+	for len(attrsVfs) != numVfs && count < 10 {
```

Comment:
nit: use common.Retry

Reply:
Same as above but you may say that if both numbers do not converge there is a real problem and may be we should abort. In that case we are in a common retry pattern with `len(attrVfs) != numVfs` considered as an error. Not sure about this one.

Reply:
The reason was that I was afraid numVfs was updated continuously as vfs were created (the reason of the asynchronous behaviour was not clear). This cannot happen (see the patch in the kernel to change the order of events, in any case (with or without) there is a single assignment of numVfs.